File: ios/Sources/DataChannel.swift
import Foundation

/// The sensor and AR data channel (docs/PROTOCOL.md §6, docs/SENSORS.md).
///
/// Channels are opt-in, and sample encoding is centralised here so the field
/// names match the spec in exactly one place.
///
/// Note the shape this is built around: sources emit *typed* samples, and each
/// subscriber encodes them itself. That is not indirection for its own sake —
/// `seq` must be contiguous per subscriber, and two subscribers asking for
/// different channel subsets cannot share one counter without one of them
/// seeing gaps, which the spec defines as dropped samples.

// MARK: - Channels

enum Channel: String, CaseIterable, Identifiable {
    case attitude
    case motion
    case arWorld = "ar.world"
    case arFace = "ar.face"
    case arPlanes = "ar.planes"
    case light
    case barometer
    case battery
    case location

    var id: String { rawValue }

    /// What the user sees in the app's channel list.
    var label: String {
        switch self {
        case .attitude: return "Orientation"
        case .motion: return "Motion & IMU"
        case .arWorld: return "AR world pose (6DoF)"
        case .arFace: return "AR face & blendshapes"
        case .arPlanes: return "AR planes"
        case .light: return "Light estimate"
        case .barometer: return "Barometer"
        case .battery: return "Battery & thermal"
        case .location: return "Location"
        }
    }

    var detail: String {
        switch self {
        case .attitude: return "Quaternion and euler angles. Cheap."
        case .motion: return "Acceleration, rotation rate, gravity, magnetometer."
        case .arWorld: return "Camera pose from ARKit. Costs battery and heat."
        case .arFace: return "52 ARKit blendshapes. Front camera only."
        case .arPlanes: return "Detected planes — a partial map of the room."
        case .light: return "Ambient intensity and colour temperature."
        case .barometer: return "Pressure and relative altitude."
        case .battery: return "Explains why the frame rate dropped."
        case .location: return "Off by default. Requires separate permission."
        }
    }

    /// Channels needing an ARKit session, which is why they are not free.
    var requiresARSession: Bool {
        self == .arWorld || self == .arFace || self == .arPlanes || self == .light
    }

    /// Face tracking uses a different ARKit configuration to world tracking.
    var requiresFaceTracking: Bool { self == .arFace }

    /// Treated separately everywhere: it never turns on implicitly.
    var isSensitive: Bool { self == .location }
}

// MARK: - Typed samples

struct AttitudeSample {
    var x: Double
    var y: Double
    var z: Double
    var w: Double
    var reference: String
    var accuracy: String?
}

struct MotionSample {
    var userAcceleration: [Float]
    var gravity: [Float]
    var rotationRate: [Float]
    var magneticField: [Float]?
    var magneticAccuracy: String?
}

struct WorldSample {
    var pose: [Float]
    var state: String
    var reason: String?
    var features: Int?
    var intrinsics: [Float]?
    var resolution: [Int]?
}

struct FaceSample {
    var tracked: Bool
    /// Only non-zero coefficients, per the spec: an absent key means zero, so
    /// sending zeros would be pure waste at 60 Hz.
    var blendShapes: [String: Float]
    var transform: [Float]?
    var leftEye: [Float]?
    var rightEye: [Float]?
    var look: [Float]?
}

struct PlaneSample {
    var id: String
    var event: String
    var center: [Float]?
    var extent: [Float]?
    var alignment: String?
    var classification: String?
}

struct LightSample {
    var lumens: Double
    var kelvin: Double?
}

struct BarometerSample {
    var kilopascals: Double
    var relativeAltitude: Double?
}

struct BatterySample {
    var level: Double
    var charging: Bool
    var thermal: String
}

struct LocationSample {
    var latitude: Double
    var longitude: Double
    var accuracy: Double
    var altitude: Double?
    var speed: Double?
    var heading: Double?
}

enum SensorSample {
    case attitude(AttitudeSample)
    case motion(MotionSample)
    case world(WorldSample)
    case face(FaceSample)
    case plane(PlaneSample)
    case light(LightSample)
    case barometer(BarometerSample)
    case battery(BatterySample)
    case location(LocationSample)

    var channel: Channel {
        switch self {
        case .attitude: return .attitude
        case .motion: return .motion
        case .world: return .arWorld
        case .face: return .arFace
        case .plane: return .arPlanes
        case .light: return .light
        case .barometer: return .barometer
        case .battery: return .battery
        case .location: return .location
        }
    }
}

// MARK: - Sample encoding

/// Encodes typed samples as NDJSON lines, numbering its own stream.
///
/// One instance per subscriber, which is what keeps `seq` contiguous.
final class SampleEncoder {
    private var sequence = 0
    private let clock: () -> Int

    /// - Parameter clock: monotonic milliseconds. Injected so tests are stable.
    init(clock: @escaping () -> Int = { MonotonicClock.milliseconds() }) {
        self.clock = clock
    }

    func reset() { sequence = 0 }

    var emitted: Int { sequence }

    func encode(_ sample: SensorSample) -> Data {
        var writer = begin(sample.channel)
        switch sample {
        case .attitude(let value):
            writer.add("q", floats: [Float(value.x), Float(value.y),
                                     Float(value.z), Float(value.w)], decimals: 6)
            let angles = Maths.euler(x: value.x, y: value.y, z: value.z, w: value.w)
            writer.add("euler", floats: [Float(angles.pitch), Float(angles.yaw),
                                         Float(angles.roll)], decimals: 3)
            writer.add("ref", value.reference)
            if let accuracy = value.accuracy { writer.add("accuracy", accuracy) }

        case .motion(let value):
            writer.add("accel", floats: value.userAcceleration)
            writer.add("gravity", floats: value.gravity)
            writer.add("rot", floats: value.rotationRate)
            if let field = value.magneticField {
                writer.add("mag", floats: field, decimals: 3)
            }
            if let accuracy = value.magneticAccuracy {
                writer.add("magAccuracy", accuracy)
            }

        case .world(let value):
            writer.add("pose", floats: value.pose)
            writer.add("state", value.state)
            if let reason = value.reason { writer.add("reason", reason) }
            if let features = value.features { writer.add("features", features) }
            if let intrinsics = value.intrinsics {
                writer.add("intrinsics", floats: intrinsics, decimals: 2)
            }
            if let resolution = value.resolution, resolution.count == 2 {
                writer.add("resolution",
                           floats: [Float(resolution[0]), Float(resolution[1])],
                           decimals: 0)
            }

        case .face(let value):
            writer.add("tracked", value.tracked)
            writer.add("blend", dictionary: value.blendShapes)
            if let transform = value.transform {
                writer.add("transform", floats: transform)
            }
            if let leftEye = value.leftEye { writer.add("leftEye", floats: leftEye) }
            if let rightEye = value.rightEye { writer.add("rightEye", floats: rightEye) }
            if let look = value.look { writer.add("look", floats: look) }

        case .plane(let value):
            writer.add("id", value.id)
            writer.add("event", value.event)
            // A removal carries only id and event; there is no geometry left.
            if value.event != "removed" {
                if let center = value.center { writer.add("center", floats: center) }
                if let extent = value.extent { writer.add("extent", floats: extent) }
                if let alignment = value.alignment { writer.add("align", alignment) }
                if let classification = value.classification {
                    writer.add("classification", classification)
                }
            }

        case .light(let value):
            writer.add("lumens", value.lumens, decimals: 2)
            if let kelvin = value.kelvin { writer.add("kelvin", kelvin, decimals: 1) }

        case .barometer(let value):
            writer.add("kpa", value.kilopascals, decimals: 4)
            if let altitude = value.relativeAltitude {
                writer.add("relAltitude", altitude, decimals: 3)
            }

        case .battery(let value):
            writer.add("level", value.level, decimals: 3)
            writer.add("charging", value.charging)
            writer.add("thermal", value.thermal)

        case .location(let value):
            writer.add("lat", value.latitude, decimals: 7)
            writer.add("lon", value.longitude, decimals: 7)
            writer.add("accuracy", value.accuracy, decimals: 2)
            if let altitude = value.altitude {
                writer.add("altitude", altitude, decimals: 2)
            }
            if let speed = value.speed { writer.add("speed", speed, decimals: 3) }
            if let heading = value.heading { writer.add("heading", heading, decimals: 2) }
        }
        return writer.finish()
    }

    /// Start a record, stamping the required t/seq/ch fields.
    private func begin(_ channel: Channel) -> JSONWriter {
        sequence += 1
        var writer = JSONWriter()
        writer.add("t", clock())
        writer.add("seq", sequence)
        writer.add("ch", channel.rawValue)
        return writer
    }
}

// MARK: - Clock

enum MonotonicClock {
    /// Monotonic milliseconds, matching the definition in PROTOCOL.md §6.3.
    ///
    /// Deliberately not wall time: the useful questions are "how far apart" and
    /// "which video frame", and both survive a clock that never jumps.
    static func milliseconds() -> Int {
        Int(ProcessInfo.processInfo.systemUptime * 1000.0)
    }
}

// MARK: - Rate limiting

/// Per-channel decimation for the `hz` query parameter.
///
/// Decimates rather than buffers: a late sample is worthless, so dropping it is
/// the correct response to a rate cap.
struct RateLimiter {
    private var lastEmitted: [String: Int] = [:]
    let intervalMillis: Int

    init(hz: Int) {
        intervalMillis = hz > 0 ? max(1, 1000 / hz) : 0
    }

    mutating func shouldEmit(channel: String, nowMillis: Int) -> Bool {
        guard intervalMillis > 0 else { return true }
        if let last = lastEmitted[channel], nowMillis - last < intervalMillis {
            return false
        }
        lastEmitted[channel] = nowMillis
        return true
    }
}

// MARK: - Subscribers

/// One `/data` subscriber: its channel filter, its rate cap, its own numbering.
final class DataSubscriber {
    let id: UUID
    private let wanted: Set<String>
    private var limiter: RateLimiter
    private let encoder: SampleEncoder
    private let send: (Data) -> Void

    init(id: UUID = UUID(), channels: Set<String>, hz: Int,
         encoder: SampleEncoder = SampleEncoder(),
         send: @escaping (Data) -> Void) {
        self.id = id
        self.wanted = channels
        self.limiter = RateLimiter(hz: hz)
        self.encoder = encoder
        self.send = send
    }

    /// An empty channel set means "everything available".
    func accepts(channel: String) -> Bool {
        wanted.isEmpty || wanted.contains(channel)
    }

    var emitted: Int { encoder.emitted }

    /// Encode and deliver, if this subscriber wants it and is not saturated.
    ///
    /// Encoding happens after the filter, so a subscriber asking for one channel
    /// costs nothing for the others.
    func offer(_ sample: SensorSample, nowMillis: Int) {
        let channel = sample.channel.rawValue
        guard accepts(channel: channel) else { return }
        guard limiter.shouldEmit(channel: channel, nowMillis: nowMillis) else { return }
        send(encoder.encode(sample))
    }
}

/// Parse the `ch` query parameter into a channel set.
func parseChannelFilter(_ raw: String?) -> Set<String> {
    guard let raw, !raw.isEmpty else { return [] }
    let names = raw.split(separator: ",").map {
        $0.trimmingCharacters(in: .whitespaces)
    }
    return Set(names.filter { !$0.isEmpty })
}

File: android/app/src/main/java/com/lostmediastudios/lostcam/DataChannel.kt
package com.lostmediastudios.lostcam

import java.util.UUID

/**
 * The data channel (docs/PROTOCOL.md §6, docs/SENSORS.md).
 *
 * Same architecture as the iOS app, and for the same reason: sources emit typed
 * samples and each subscriber encodes them itself, because `seq` must be
 * contiguous per subscriber and two subscribers with different channel filters
 * cannot share a counter without one of them seeing gaps — which the spec defines
 * as dropped samples.
 *
 * What Android does *not* offer is as important as what it does. The `ar.*`
 * channels are absent in this version: ARCore's Augmented Faces API produces a
 * mesh and region poses but no blendshape coefficients, so an `ar.face` channel
 * here could never mean what it means on iOS, and inventing coefficients would be
 * worse than absence. `/info` advertises only what is real.
 */
enum class Channel(val wireName: String, val label: String, val detail: String) {
    ATTITUDE("attitude", "Orientation", "Quaternion and euler angles. Cheap."),
    MOTION("motion", "Motion & IMU", "Acceleration, rotation rate, gravity, magnetometer."),
    BAROMETER("barometer", "Barometer", "Pressure and relative altitude."),
    BATTERY("battery", "Battery & thermal", "Explains why the frame rate dropped."),
    LOCATION("location", "Location", "Off by default. Needs its own permission."),
    ;

    /** Treated separately everywhere; never enabled implicitly. */
    val isSensitive: Boolean get() = this == LOCATION

    companion object {
        fun fromWireName(name: String): Channel? =
            entries.firstOrNull { it.wireName == name }
    }
}

// MARK: - Typed samples

sealed interface SensorSample {
    val channel: Channel

    data class Attitude(
        val x: Double,
        val y: Double,
        val z: Double,
        val w: Double,
        val reference: String,
        val accuracy: String?,
    ) : SensorSample {
        override val channel get() = Channel.ATTITUDE
    }

    data class Motion(
        /** In g, not m/s^2 — see Maths.toG. */
        val userAcceleration: FloatArray,
        val gravity: FloatArray,
        val rotationRate: FloatArray,
        val magneticField: FloatArray?,
        val magneticAccuracy: String?,
    ) : SensorSample {
        override val channel get() = Channel.MOTION
    }

    data class Barometer(
        val kilopascals: Double,
        val relativeAltitude: Double?,
    ) : SensorSample {
        override val channel get() = Channel.BAROMETER
    }

    data class Battery(
        val level: Double,
        val charging: Boolean,
        val thermal: String,
    ) : SensorSample {
        override val channel get() = Channel.BATTERY
    }

    data class Location(
        val latitude: Double,
        val longitude: Double,
        val accuracy: Double,
        val altitude: Double?,
        val speed: Double?,
        val heading: Double?,
    ) : SensorSample {
        override val channel get() = Channel.LOCATION
    }
}

// MARK: - Encoding

/**
 * Encodes typed samples as NDJSON lines, numbering its own stream.
 * One instance per subscriber, which is what keeps `seq` contiguous.
 */
class SampleEncoder(private val clock: () -> Long = { MonotonicClock.millis() }) {
    private var sequence = 0

    val emitted: Int get() = sequence

    fun reset() {
        sequence = 0
    }

    fun encode(sample: SensorSample): ByteArray {
        val writer = begin(sample.channel)
        when (sample) {
            is SensorSample.Attitude -> {
                writer.addFloats(
                    "q",
                    floatArrayOf(
                        sample.x.toFloat(), sample.y.toFloat(),
                        sample.z.toFloat(), sample.w.toFloat(),
                    ),
                    decimals = 6,
                )
                val angles = Maths.euler(sample.x, sample.y, sample.z, sample.w)
                writer.addFloats(
                    "euler",
                    floatArrayOf(
                        angles[0].toFloat(), angles[1].toFloat(), angles[2].toFloat(),
                    ),
                    decimals = 3,
                )
                writer.add("ref", sample.reference)
                sample.accuracy?.let { writer.add("accuracy", it) }
            }

            is SensorSample.Motion -> {
                writer.addFloats("accel", sample.userAcceleration)
                writer.addFloats("gravity", sample.gravity)
                writer.addFloats("rot", sample.rotationRate)
                sample.magneticField?.let { writer.addFloats("mag", it, decimals = 3) }
                sample.magneticAccuracy?.let { writer.add("magAccuracy", it) }
            }

            is SensorSample.Barometer -> {
                writer.add("kpa", sample.kilopascals, decimals = 4)
                sample.relativeAltitude?.let {
                    writer.add("relAltitude", it, decimals = 3)
                }
            }

            is SensorSample.Battery -> {
                writer.add("level", sample.level, decimals = 3)
                writer.add("charging", sample.charging)
                writer.add("thermal", sample.thermal)
            }

            is SensorSample.Location -> {
                writer.add("lat", sample.latitude, decimals = 7)
                writer.add("lon", sample.longitude, decimals = 7)
                writer.add("accuracy", sample.accuracy, decimals = 2)
                sample.altitude?.let { writer.add("altitude", it, decimals = 2) }
                sample.speed?.let { writer.add("speed", it, decimals = 3) }
                sample.heading?.let { writer.add("heading", it, decimals = 2) }
            }
        }
        return writer.finish()
    }

    /** Start a record, stamping the required t/seq/ch fields. */
    private fun begin(channel: Channel): JsonWriter {
        sequence += 1
        return JsonWriter()
            .add("t", clock())
            .add("seq", sequence)
            .add("ch", channel.wireName)
    }
}

object MonotonicClock {
    /**
     * Monotonic milliseconds, matching PROTOCOL.md §6.3. `nanoTime` rather than
     * `currentTimeMillis` deliberately: the useful questions are "how far apart"
     * and "which video frame", and both survive a clock that never jumps.
     */
    fun millis(): Long = System.nanoTime() / 1_000_000L
}

// MARK: - Rate limiting

/**
 * Per-channel decimation for the `hz` query parameter. Decimates rather than
 * buffers: a late sample is worthless, so dropping it is the right response.
 */
class RateLimiter(hz: Int) {
    val intervalMillis: Long = if (hz > 0) maxOf(1L, 1000L / hz) else 0L
    private val lastEmitted = mutableMapOf<String, Long>()

    fun shouldEmit(channel: String, nowMillis: Long): Boolean {
        if (intervalMillis <= 0) return true
        val last = lastEmitted[channel]
        if (last != null && nowMillis - last < intervalMillis) return false
        lastEmitted[channel] = nowMillis
        return true
    }
}

// MARK: - Subscribers

/** One `/data` subscriber: channel filter, rate cap, and its own numbering. */
class DataSubscriber(
    private val wanted: Set<String>,
    hz: Int,
    private val encoder: SampleEncoder = SampleEncoder(),
    private val send: (ByteArray) -> Unit,
) {
    val id: UUID = UUID.randomUUID()
    private val limiter = RateLimiter(hz)

    /** An empty channel set means "everything available". */
    fun accepts(channel: String): Boolean = wanted.isEmpty() || wanted.contains(channel)

    val emitted: Int get() = encoder.emitted

    /**
     * Encode and deliver, if this subscriber wants it and is not saturated.
     * Encoding happens after the filter, so asking for one channel costs nothing
     * for the others.
     */
    fun offer(sample: SensorSample, nowMillis: Long) {
        val channel = sample.channel.wireName
        if (!accepts(channel)) return
        if (!limiter.shouldEmit(channel, nowMillis)) return
        send(encoder.encode(sample))
    }
}

/** Parse the `ch` query parameter into a channel set. */
fun parseChannelFilter(raw: String?): Set<String> {
    if (raw.isNullOrEmpty()) return emptySet()
    return raw.split(",").map { it.trim() }.filter { it.isNotEmpty() }.toSet()
}
